File: compiler/bench/maze.fs
\ expected: 1000
\ Maze solver using flood fill

16 constant SIZE
create maze SIZE SIZE * allot
create visited SIZE SIZE * allot

: m@ ( x y -- val ) SIZE * + maze + c@ ;
: m! ( val x y -- ) SIZE * + maze + c! ;
: v@ ( x y -- val ) SIZE * + visited + c@ ;
: v! ( val x y -- ) SIZE * + visited + c! ;

: in-bounds? ( x y -- flag )
  dup 0>= swap SIZE < and
  swap dup 0>= swap SIZE < and and ;

: init-maze ( -- )
  \ Create simple maze with path from (0,0) to (15,15)
  SIZE SIZE * 0 do 0 maze i + c! loop
  SIZE SIZE * 0 do 0 visited i + c! loop
  \ Clear path along edges and diagonal
  SIZE 0 do 1 i 0 m! loop        \ top row
  SIZE 0 do 1 SIZE 1- i m! loop  \ bottom row
  SIZE 0 do 1 0 i m! loop        \ left column
  SIZE 0 do 1 i SIZE 1- m! loop  \ right column
  ;

variable found

: solve ( x y goalx goaly -- found? )
  2over 2over = swap 3 pick = and if
    2drop 2drop 1 exit
  then
  2over in-bounds? 0= if 2drop 2drop 0 exit then
  2over m@ 0= if 2drop 2drop 0 exit then
  2over v@ if 2drop 2drop 0 exit then
  2over 1 rot rot v!
  \ Try all 4 directions
  2over 1+ 2over recurse if 2drop 2drop 1 exit then
  2over 1- 2over recurse if 2drop 2drop 1 exit then
  2over swap 1+ swap 2over recurse if 2drop 2drop 1 exit then
  2over swap 1- swap 2over recurse if 2drop 2drop 1 exit then
  2drop 2drop 0 ;

: run-maze ( -- found? )
  SIZE SIZE * 0 do 0 visited i + c! loop
  0 0 SIZE 1- SIZE 1- solve ;

: main
  0
  1000 0 do
    init-maze
    run-maze if 1+ then
  loop
  . cr ;

File: compiler/bench/maze.c
// expected: 1000
// Maze solver using flood fill
#include <stdio.h>
#include <string.h>

#define SIZE 16

char maze[SIZE * SIZE];
char visited[SIZE * SIZE];

int m(int x, int y) { return maze[y * SIZE + x]; }
void ms(int val, int x, int y) { maze[y * SIZE + x] = val; }
int v(int x, int y) { return visited[y * SIZE + x]; }
void vs(int val, int x, int y) { visited[y * SIZE + x] = val; }

int in_bounds(int x, int y) {
    return x >= 0 && x < SIZE && y >= 0 && y < SIZE;
}

void init_maze(void) {
    memset(maze, 0, SIZE * SIZE);
    memset(visited, 0, SIZE * SIZE);
    for (int i = 0; i < SIZE; i++) ms(1, i, 0);
    for (int i = 0; i < SIZE; i++) ms(1, i, SIZE - 1);
    for (int i = 0; i < SIZE; i++) ms(1, 0, i);
    for (int i = 0; i < SIZE; i++) ms(1, SIZE - 1, i);
}

int solve(int x, int y, int gx, int gy) {
    if (x == gx && y == gy) return 1;
    if (!in_bounds(x, y)) return 0;
    if (!m(x, y)) return 0;
    if (v(x, y)) return 0;
    vs(1, x, y);
    if (solve(x + 1, y, gx, gy)) return 1;
    if (solve(x - 1, y, gx, gy)) return 1;
    if (solve(x, y + 1, gx, gy)) return 1;
    if (solve(x, y - 1, gx, gy)) return 1;
    return 0;
}

int run_maze(void) {
    memset(visited, 0, SIZE * SIZE);
    return solve(0, 0, SIZE - 1, SIZE - 1);
}

int main(void) {
    int count = 0;
    for (int i = 0; i < 1000; i++) {
        init_maze();
        if (run_maze()) count++;
    }
    printf("%d\n", count);
    return 0;
}
